Add OpenTelemetry sampling conventions

jmacd opened the pull request:
## Changes

Introduces 3 conventional attributes to describe sampling in OpenTelemetry collection pipelines. This specification refers to [OTEP 235](https://github.com/open-telemetry/oteps/blob/main/text/trace/0235-sampling-threshold-in-trace-state.md).

Related to the specification change (from OTEP 235) in https://github.com/open-telemetry/opentelemetry-specification/pull/3910.

Related to the specification of about representing trace context in logs in https://github.com/open-telemetry/opentelemetry-specification/pull/3909.

Prototype for a Collector sampler based on these attributes in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29720 and [WIP] https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24811.

Part of https://github.com/open-telemetry/opentelemetry-specification/issues/1413

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

lmolkova commented on model/otel/sampling.yaml:
```diff
@@ -0,0 +1,49 @@
+  - id: sampling.common
+    type: attribute_group
+    brief: 'Semantic conventions describing sampling in OpenTelemetry.'
+    prefix: sampling
```
Since it's otel-related attribute, we should consider putting it in the `otel.` namespace

jmacd replied:
I've added text to argue in favor of "sampling" because this "sampling.priority" is already defined.

lmolkova commented on model/otel/sampling.yaml:
```diff
@@ -0,0 +1,49 @@
+    brief: 'Semantic conventions describing sampling in OpenTelemetry.'
+    prefix: sampling
+    attributes:
+      - id: priority
```
we try to put all attributes into the registry - please use `url.full` as an example.
Then individual conventions can refer to that attribute

jmacd replied:
Done!

lmolkova commented on model/otel/sampling.yaml:
```diff
@@ -0,0 +1,49 @@
+          collected.  This convention derives from a conventional tag
+          with the same name in the OpenTracing specification.
+        examples: [1, 0]
+        tag: otel-span-attributes
```
is it necessary ?

jmacd replied:
Removed. Now there is one table of attributes, besides which @kentquirk thinks the attributes should be useful for spans, too.

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+
+### Sampling threshold
+
+The OpenTelemetry sampling decision is defined in terms in terms of a
```
```suggestion
The OpenTelemetry sampling decision is defined in terms of a
```

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+R`.  An item is sampled when the item's sampling threshold is less or
+equal to its randomness.
+
+When determining sampling randomness from a span, implementations
```
`sampling.randomness` attribute is defined below, but here we say it's coming from threshold or tracestate - is that right?

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+
+### Sampling randomness
+
+When determining sampling randomness from a span, implementations
```
what implementation are you referring to here?

lmolkova replied:
also this is duplication of line 36 and next few ones

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+- use the `sampling.randomness` attribute value if it is present, or
+- use the least significant 56 bits of the W3C Trace Context TraceID, as described in the W3C Trace Context Level 2 specification.
+
+Callers SHOULD NOT require trace flags to have the Trace Context Level
```
which callers are you referring to here? Callers of semantic conventions?

jmacd replied:
Sorry! I removed "Callers". This is a question concerning parent-based and tail-samplers. I replaced this with "Sampler implementations" as well.

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+
+## Span attributes
+
+The following attributes are recognized for Spans.  Since probability
```
we don't really divide attributes between signals anymore, so I suggest to merge span and log attributes.
 
If randomness and threshold don't appear on spans, it can be reflected via something like `requirement_level: conditionally_required: on log records`  or notes on corresponding attribute.

jmacd replied:
I think I've fixed this. However, I opted to remove the `tag` and leave out `requirement_level` for now, until seeing more feedback. I think these are probably best explained as conditionally required when opting-in to sampling, on log records (maybe when not using trace ID randomness). I've added normative text to explain this.

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+<!-- toc -->
+<!-- tocstop -->
+
+These attributes reflect the effect of sampling in a telemetry
```
which component is expected to follow this semconv and set these attributes? A sampler? User application which wants to customize sampler behavior? 


also, who's the consumer is this data? other pipeline components? samplers? backends? end users?

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+<!-- semconv sampling.common(full,tag=otel-span-attributes) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `sampling.priority` | int | Enables prioritizing and de-prioritizing collection of telemetry. [1] | `1`; `0` | Recommended |
```
Is it possible to describe the parameter from the user perspective? E.g. "Describes a sampler decision to prioritize or deprioritize collection of this telemetry item"

jmacd replied:
I changed this to "Allows users to prioritize and de-prioritize items of telemetry." and added a section on "Overriding sampling decisions".

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+|---|---|---|---|---|
+| `sampling.priority` | int | Enables prioritizing and de-prioritizing collection of telemetry. [1] | `1`; `0` | Recommended |
+
+**[1]:** This attribute is useful to express the user's desire to collect telemetry, despite the sampling configuration in place.  Non-zero values MAY prioritize the item for collection independent of probability sampling.  Zero values have the reverse effect and MAY cause telemetry not to be collected.  This convention derives from a conventional tag with the same name in the OpenTracing specifi ...
```
if only 0 and 1 are supported, please consider converting it into enum and adding 0 and 1 as members.

jmacd replied:
Actually, this is described as an integer to allow users to have multiple priorities--and this convention is the one inherited from OpenTracing that we can't really change (e.g., we expect there are Jaeger users actively using this attribute). The only implementation inside OpenTelemetry that I'm aware of uses a zero vs non-zero test ([probabilisticsamplingprocessor](https://pkg.go.dev/go.opentelemetry.io/collector/processor/probabilisticsamplerprocessor)).

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+|---|---|---|---|---|
+| `sampling.priority` | int | Enables prioritizing and de-prioritizing collection of telemetry. [1] | `1`; `0` | Recommended |
+
+**[1]:** This attribute is useful to express the user's desire to collect telemetry, despite the sampling configuration in place.  Non-zero values MAY prioritize the item for collection independent of probability sampling.  Zero values have the reverse effect and MAY cause telemetry not to be collected.  This convention derives from a conventional tag with the same name in the OpenTracing specifi ...
```
Also, is it helpful to refer to OpenTracing here? Perhaps reference to the OTEP would be more relevant?

jmacd replied:
I've quoted the OpenTracing specification in a normative section and removed it from this location. This attribute has a historical, vague and fixed definition.

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `sampling.priority` | int | Enables prioritizing and de-prioritizing collection of telemetry. [1] | `1`; `0` | Recommended |
+| `sampling.randomness` | string | The source of randomness for making probability sampling decisions, when it is not otherwise recorded. [2] | `ce929d0e0e4736` | Recommended |
```
note mentions it's optional, should requirement_level be opt_in?

jmacd replied:
I've written text to explain when this should be used, which is only for logs, and only when not using trace_id randomness (or for a couple other reasons mentioned in the text).

lmolkova commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+|---|---|---|---|---|
+| `sampling.priority` | int | Enables prioritizing and de-prioritizing collection of telemetry. [1] | `1`; `0` | Recommended |
+| `sampling.randomness` | string | The source of randomness for making probability sampling decisions, when it is not otherwise recorded. [2] | `ce929d0e0e4736` | Recommended |
+| `sampling.threshold` | string | Sampling probability as specified by OpenTelemetry. [3] | `c`; `ff8` | Recommended |
```
Is it a static value? Do we really want to have this attribute on every log record if it is or maybe it should be a resource or instrumentation scope attribute?

kentquirk replied:
It is very much not a static value and can vary with each log record, so cannot be a higher-level attribute.

jmacd replied:
Moreover, as I've now written in the text, we can't in today's conventions safely apply a resource-level or scope-level attribute for Threshold without potentially overcounting due to spans that were sampled because of `sampling.priority`. In earlier (experimental) specifications, we had a T-value expression equivalent to "zero count", which might be a solution. For now, the text says we aren't defining these attributes for resources or scopes.

pyohannes commented on docs/otel/sampling.md:
```diff
@@ -0,0 +1,101 @@
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `sampling.priority` | int | Enables prioritizing and de-prioritizing collection of telemetry. [1] | `1`; `0` | Recommended |
+| `sampling.randomness` | string | The source of randomness for making probability sampling decisions, when it is not otherwise recorded. [2] | `ce929d0e0e4736` | Recommended |
```
Shouldn't `sampling.randomness` and `sampling.threshold` be (also) on spans, and not only on logs?

kentquirk replied:
Yes, they should.

jmacd replied:
I would expect for spans that these are encoded in the `tracestate`, but I don't have a problem with accepting them on spans for other situations. In the text, now, I've stated that only one form should appear in any given item. 

@kentquirk why wouldn't you use the tracestate in this case? I'm concerned about situations where there are both a tracestate and an attribute, for example, or whether you can mix a `sampling.threshold` attribute with a `tracestate`-derived R-value.

jmacd commented:
@lmolkova @pyohannes Thank you! I realized from your responses that this repository needs to contain more detail, whereas I had been planning on rolling that detail into https://github.com/open-telemetry/opentelemetry-specification/pull/3910 and now I've put some of that text into this PR.

I created a file in the attributes registry and moved the attribute definitions there.

I am hoping to justify the use of "sampling" as a prefix, for historical reasons. I removed the `tag` and decided to wait for more feedback before introducing a requirements-level statement, because these attributes are somewhat unusual.

Since spans record `tracestate`, unlike log records, we are proposing two new attributes that are only useful for log records corresponding with fields that are defined for use in `tracestate`. I had been using the `tag` to de-select those two attributes in the span attributes and leave them in the logs attributes. Now, there is just one section called "Sampling attributes" and the docs explain this.

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+| `sampling.randomness` | string | The source of randomness for making probability sampling decisions, when it is not otherwise recorded. [2] | `ce929d0e0e4736` |
+| `sampling.threshold` | string | Sampling probability as specified by OpenTelemetry. [3] | `c`; `ff8` |
+
+**[1]:** This attribute is useful to express a user's desire to collect an item of telemetry, in spite of the default sampling mechanism.  Samplers SHOULD export items with a non-zero value, independent of probability sampling. Samplers SHOULD NOT export items with this attribute set to zero.
```
Samplers or exporters?
Also, if we expect samplers to do something about this attribute, it'd be great to clarify that it applies to tail-based sampling

jmacd replied:
I've added `sampling_relevant: true`. There are normative statements about how this should be used in samplers, and I expect them to be used in both head- and tail-based scenarios. The collector's `probabilisticsamplerprocessor` uses this field as specified to bypass its intermediate sampling decisions.

jmacd replied:
I added an example tail-sampler scenario explaining how these attributes might be synthesized, next to the current head-sampler examples.

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+sampling described here.  In that sense, producers and consumers
+should see these attributes as telemetry metadata.
+
+The attributes described here use non-standard naming conventions.
```
thank you for the explanation, I think we don't need to add it to the doc itself. Also I'd still suggest removing references to OpenTracing - they provide great historical context, but don't seem practical

jmacd replied:
Removed it, thanks!

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+### Sampling threshold
+
+When determining the Threshold value from an item of telemetry,
+implementations SHOULD:
```
I still don't understand implementations of what. Sampler? Processor? Exporter? Collector/etc pipeline component?

jmacd replied:
In the prototype for probabilistictailsampler, https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24811, the processor is an "intermediate span sampler" which may both read the threshold and write a new threshold after making a new sampling decision. So, we expect this to be used by head samplers, tail samplers, and backends for constructing metrics from sampled items. I also wonder if the changes in https://github.com/open-telemetry/opentelemetry-specification/pull/3910 help answer this sort of question?

jmacd replied:
I added a tail-sampler example here: [4e8870d](https://github.com/open-telemetry/semantic-conventions/pull/793/commits/4e8870d74da0cc35f80737f7800a096a4000c15d)

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+
+Instrumentation authors and end-users that wish to prioritize an item
+of telemetry for collection in spite of sampling can add a
+`sampling.priority` attribute.  This attribute is inherited as a
```
Are we adding it  for the sake of compatibility with OpenTracing? It does not seem so and I'd prefer to avoid "inhering as legacy" part and removing references to the historical context.

jmacd replied:
Removed & done. Thanks.

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+legacy from the OpenTracing project, which gave the following
+definition.
+
+> If greater than 0, a hint to the Tracer to do its best to capture
```
Can we add this to the brief/note on the attribute instead?

jmacd replied:
Done.

lmolkova commented on model/registry/sampling.yaml:
```diff
@@ -0,0 +1,47 @@
+    attributes:
+      - id: priority
+        type: int
+        brief: Enables prioritizing and de-prioritizing collection of telemetry.
```
```suggestion
        brief: Allows users and instrumentations to prioritize collection of reported telemetry item
```

jmacd replied:
Applied.

lmolkova commented on model/registry/sampling.yaml:
```diff
@@ -0,0 +1,47 @@
+          sampling mechanism.  Samplers SHOULD export items with a
+          non-zero value, independent of probability sampling.
+          Samplers SHOULD NOT export items with this attribute set to
+          zero.
```
```suggestion
        If greater than 0, a hint to the Tracer to do its best to capture
        the trace. If 0, a hint to the Tracer to not-capture the trace. If
        absent, the Tracer should use its default sampling mechanism.
```

jmacd replied:
Done.

lmolkova reviewed (commented):
Thanks a lot for the additional context, I still have some questions and suggestions

jmacd commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+100% sampling.
+
+When Threshold is not provided, no information about probability
+sampling is available and
```
Fixed!

jmacd commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+These attributes MAY be modified by components in a collection
+pipeline to convey successive sampling that has been carried out for a
+particular item of telemetry, using the conventions for consistent
+sampling described here.  In that sense, producers and consumers
```
Applied.

kentquirk commented on .chloggen/793.yaml:
```diff
@@ -0,0 +1,22 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Adds `sampling.priority` from OpenTracing. Adds `sampling.randomness` and `sampling.threshold` from [OTEP 235](https://github.com/open-telemetry/oteps/blob/main/text/trace/0235-sampling-threshold-in-trace-state.md).
```
```suggestion
subtext: >
Adds `sampling.priority` from OpenTracing. 
Adds `sampling.randomness` and `sampling.threshold` from [OTEP 235](https://github.com/open-telemetry/oteps/blob/main/text/trace/0235-sampling-threshold-in-trace-state.md).
```

jmacd replied:
I accept -- whether yamllint accepts or not!

kentquirk reviewed (approved):
I'm very happy about this document.

jmacd commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,242 @@
+is included in the Span data model, it is not common to express
+sampling randomness and threshold using attributes.
+
+<!-- semconv registry.sampling(omit_requirement_level) -->
```
Done! I think I had misunderstood the data model slightly. Now, I have "individual" sampling conventions for logs and traces, which refer to the attribute registry. Traces refer to 1 attribute, logs refers to 3 attributes, which are `opt_in` and `conditionally_required`.

kalyanaj commented on docs/sampling/README.md:
````diff
@@ -0,0 +1,282 @@
+```
+   if err := doSomething(); err != nil {
+     if err == VERY_SERIOUS {
+       span.SetAttribute("sampling.priority", 10000)
````
Nit: Is the idea that even if an original sampling decision was to drop, you can override it by setting this attribute? If so, IMHO, the name "priority" doesn't directly convey the override aspect.  Hence, depending on what use cases we want to allow (change decision from drop -> keep, or change decision from keep -> drop, or both), I wonder if this should be a simple bool attribute such as "sampling.keep" or "sampling.include" or "sampling.override".

In the future, if we expect pipelines to prioritize between multiple levels (e.g., if I could keep only X%, prefer the ones with the higher priority), then we could still add a new attribute to convey that relative priority.

Thoughts?

jmacd replied:
@kalyanaj I tried to address your concern in https://github.com/open-telemetry/semantic-conventions/pull/793#issuecomment-1985953857.

@lmolkova Yes. Another way of saying this (which OTEP 235 explains) is that Thresholds cannot decrease as a result of sampling, but they can increase. OTEP 235 writes

>  In other words, th MUST NOT be decreased (as it is not possible to retroactively adjust an earlier stage's sampling probability), and it MUST be increased if a lower sampling probability was used.

The trace SDK specification confines what we can do for SDKs, in the sense that there is not Sampler decision corresponding with not sampling, but recording the span and possibly changing the decision to export the unsampled span anyway. See e.g.,

https://github.com/open-telemetry/opentelemetry-specification/issues/2918
https://github.com/open-telemetry/opentelemetry-specification/issues/2986

kalyanaj commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,282 @@
+
+A log record containing a TraceID is received with no sampling
+attributes and is selected by a tail sampler at 10% probability.  A
+sampling threshold is inserted
```
```suggestion
sampling threshold is inserted:
```

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,282 @@
+### Overriding sampling randomness
+
+Sampling system designers are able to override sampling randomness on
+an item-by-item basis, which may be done for several reasons, including
```
nit: It seems item-by-item basis is only available for logs since tracestate will likely to be propagated to children and out-of-process.

jmacd replied:
Agreed -- see a55667f.

lmolkova commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,282 @@
+When determining the Randomness value from an item of telemetry,
+sampler implementations SHOULD:
+
+- use the `tracestate` OpenTelemetry R-value field (`rv`) if it is present (spans only), or
```
tracestate maybe present and accessible within the process through current span and I assume it's desirable for spans and nested logs to have the same value, should there be any coordination between tracestate from span context and the attribute value in this case?

jmacd replied:
Agreed -- see 79275a3.

jmacd commented:
@kalyanaj considering the `sampling.priority` attribute, some background may help us. This attribute was originally specified in OpenTracing and appears verbatim in this PR. There is certainly a suggestion that higher priorities receive more consideration, but it does not include an algorithm or mathematical specification. I understand that this is still in use for instrumentation derived from OpenTracing, and there are several reasons I included it (and Datadog also has this semantic, presumably from OpenTracing).

- I wanted to justify use of the `sampling.` namespace by incorporating the old specification
- The [sampler processor I have modified for the prototype](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24811) [implements this attribute as a bypass](https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/10438)

In the past we've entertained the idea of "Zero adjusted count" -- an explicit T-value that says an item has not been probability sampled (the corresponding value 0x100000000000000 reflects this adequately, i.e., the 100%-rejection threshold, but is out of range in this specification). I see `sampling.priority` as a weak substitute for zero adjusted count. If I have a span that passes through a probability sampler because of `sampling.priority`, no T-value will be applied. If no T-value was applied, I  can't accurately count the item, but there will be two cases:

- No T-value, no sampling.priority -- definitely unknown
- No T-value, with sampling.priority -- probably bypassed a sampler

However, this is a weak substitute. I would be interested in adding a tracestate field to signify "unknown adjusted count", but I think the sampling.priority attribute should remain either way.

lmolkova reviewed (commented):
I think this is a great improvement for sampling story. I have some general non-blocking (at least for the initial version) concerns:

1. Sampling attributes (as explained in this PR) are effectively a metadata used by telemetry pipeline components. They are useful when aggregating sampled data, but not in general user queries. We almost want them to be dropped inside the pipeline and never exported to the final destination after all transformations have ben completed.
    - I think this concern can be mitigated by putting them into some namespace that emphasizes this. I still think `otel.` might be a good place for them
2. The inconsistency between spans and logs (tracestate vs attributes). Should nested logs inherit sampling properties from parent span's tracestate? How do I set randomness for one span, but not its children? What if I want to set randomness for a group of spans in the instrumentation library - I should never touch tracestate there.

jmacd commented on docs/attributes-registry/sampling.md:
```diff
@@ -0,0 +1,27 @@
+| `sampling.randomness` | string | The source of randomness for making probability sampling decisions, when it is not otherwise recorded. [2] | `ce929d0e0e4736` |
+| `sampling.threshold` | string | Sampling probability as specified by OpenTelemetry. [3] | `c`; `ff8` |
+
+**[1]:** If greater than 0, a hint to the Tracer to do its best to capture the trace. If 0, a hint to the Tracer to not-capture the trace. If absent, the Tracer should use its default sampling mechanism.
```
See a28f48e.

jmacd commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,282 @@
+
+Samplers and sampling processors SHOULD pass items of telemetry to the
+exporter, independent of their default sampling mechanism, when the
+`sampling.priority` attribute is present and non-zero.  Although the
```
Agreed -- See a28f48e

jmacd commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,282 @@
+| [`sampling.randomness`](../attributes-registry/sampling.md) | string | The source of randomness for making probability sampling decisions, when it is not otherwise recorded. [2] | `ce929d0e0e4736` | Conditionally Required: [3] |
+| [`sampling.threshold`](../attributes-registry/sampling.md) | string | Sampling probability as specified by OpenTelemetry. [4] | `c`; `ff8` | Conditionally Required: [5] |
+
+**[1]:** If greater than 0, a hint to the Tracer to do its best to capture the trace. If 0, a hint to the Tracer to not-capture the trace. If absent, the Tracer should use its default sampling mechanism.
```
Agreed -- see c3650d2.

jmacd commented:
>  I think this concern can be mitigated by putting them into some namespace that emphasizes this. I still think otel. might be a good place for them

Or we could go the other direction, and define `logstate` for expressing equivalent information as `tracestate` contains. I don't particularly like expressing metadata via attributes either, but it is a path with not much resistance.  OTOH, Metric instruments now have metadata, so maybe we could add logs metadata as a separate field. See https://github.com/open-telemetry/opentelemetry-proto/blob/9d139c87b52669a3e2825b835dd828b57a455a55/opentelemetry/proto/metrics/v1/metrics.proto#L199.

> The inconsistency between spans and logs (tracestate vs attributes). Should nested logs inherit sampling properties from parent span's tracestate? 

They should use the tracestate for randomness when it is available. I changed the text on this, since you asked.

> How do I set randomness for one span, but not its children? 

I've added "the R-value should not be modified, once set, in a Context", because I do not believe you should do this. If you really must, then I believe it is permitted in the Trace API specification, which writes:

```
Such changes then make sense only right before [`SpanContext` propagation](../context/api-propagators.md) or [telemetry data exporting](sdk.md#span-exporter).
In both cases, `Propagator`s and `SpanExporter`s may create a modified `TraceState` copy before serializing it to the wire.
```

(As a tangent, I believe the specification permits changing the SpanContext between parent and child spans where there is no propagator involved, e.g., the Go SDK has a `trace.ContextWithSpanContext()` method.)

> What if I want to set randomness for a group of spans in the instrumentation library - I should never touch tracestate there.

It makes the most sense to set R-value in a root span and let it stay constant through a trace. If you have a group of root spans, then this is exactly one of the documented scenarios. You can give the same R-value to more than one root span and they will sample or not-sample multiple correlated traces consistently.

I don't think the semantic conventions need to dictate how the convention is used so carefully. I think it's OK to modify the tracestate and set an R-value in a non-root location, it just means there will be broken traces under some conditions. Some of the questions raised here may belong as statements in the corresponding specification PR, https://github.com/open-telemetry/opentelemetry-specification/pull/3910.

kentquirk commented:
> 1. Sampling attributes (as explained in this PR) are effectively a metadata used by telemetry pipeline components. They are useful when aggregating sampled data, but not in general user queries. We almost want them to be dropped inside the pipeline and never exported to the final destination after all transformations were completed.

I strongly disagree -- when analyzing telemetry, having the sampling attributes available can very much help in the analysis. As a simple example, imagine that your telemetry is set up to sample most data at 1% but keep all errors. If you are comparing the number of errors to the total telemetry volume, you'll be wrong by a factor of 100!

jmacd commented:
@kentquirk @PeterF778 @oertl @kalyanaj For the record --

As discussed in Sampling SIG last week, the major question posed in this PR is whether to use a new metadata-specific field for recording the Logs equivalent of `tracestate`, vs (as proposed here) to fall back on the use of generic attributes to signal sampling conventions for log records. Conditions are mixed across signals currently -- traces have `tracestate`, metrics now have `metadata`, and logs are missing a similar construct. This question is open to the community.

There is also a more subtle semantics question raised. I've discovered a surprise in the existing implementation of the `probabilisticsamplerprocessor`, which has a very different interpretation of the "sampling priority" concept **when applied to logs**. This is not a _direct_ use of the attribute named `sampling.priority` but a configurable use w/ a field named "sampling_priority" that indicates the sampling priority attribute name for log records. So it's approximately the same thing, but the implementation is quite different.

The logging processor treats the "sampling priority" as a request to be sampled, indicated in percent. So a logging sampling priority value of 100 in that code means "always sample", whereas in the same code for traces the `sampling.priority` value of 1 means the thing inherited from OpenTracing, quoted in this PR.  Please consider ways to resolve this difference with me. 

Attempting to re-implement the same behavior in this sampler and respect the logging priority semantic broke a few assumptions, but those assumptions were already challenged in https://github.com/open-telemetry/oteps/pull/250. We can stop thinking of "sampling priority" as a way to bypass the default mechanism, and reinterpret it as an implicit `AnyOf` composition, where a special-purpose sampler is composed that returns a threshold somehow stated in the priority attribute. For this logs implementation, then, sampling priority of 100 means returning threshold `0`, sampling priority 50 means returning threshold `8`, and so on. This sampler returns "do not sample" when there is no attribute, bypassing itself. I guess the proper name for a feature like this is `sampling.requested_at_percent` (i.e., it doesn't mean "I was sampled", it means "please sample me by so much")

Another direction, which may deserve attention, is to treat this "priority" concept, as a weight multiplier, and then to use a weighted sampling algorithm. (E.g., I might recommend https://github.com/lightstep/varopt.) Items without the attribute are given weight 1, and otherwise items are entered into a reservoir their stated priority weight. I think this gives us a more natural definition for "priority", and I think this is the spirit in which the OpenTracing specification was written. An item with sampling priority 10 should be treated as 10 times more valuable than ordinary items. If every item has this exact attribute, it becomes meaningless. Otherwise, items with priority 10 are more likely to be selected, depending on population size. The problem with this approach, is that it is an intentional bias. Logs-to-metrics pipelines would therefore count those logs in a non-representative way, unless we're careful and specify a solution.

jmacd commented:
Reviewers: I see no reason to continue promoting `sampling.priority` as a convention, despite its definition in OpenTracing. I found a very different definition for a near-identical concept in the OpenTelemetry contrib-collector probabilistic sampler processor for logs. If we used the logs processor definition, "sampling.priority" would equal a request to sample at a particular percentage, where there is an implicit conjunction. In the terminology of [OTEP 250](https://github.com/open-telemetry/oteps/pull/250), the priority mechanism is the FIRST in the conjunction (making it possible to drop before evaluating the second).

Therefore, I have revised this PR only to specify `sampling.threshold` and `sampling.randomness`.

jmacd commented on model/trace/sampling.yaml:
```diff
@@ -0,0 +1,7 @@
+    type: attribute_group
+    brief: 'Semantic convention describing trace attributes related to sampling.'
+    attributes:
+      - ref: sampling.priority
```
Fixed.

jmacd commented:
Updated: `sampling.priority` is really removed, now.
This leaves the question in https://github.com/open-telemetry/semantic-conventions/pull/793#pullrequestreview-1925797614 (and https://github.com/open-telemetry/semantic-conventions/pull/793#issuecomment-2004318492) about whether we agree to use log record attributes to convey metadata about the collection path.

PeterF778 commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,262 @@
+The OpenTelemetry sampling decision is defined in terms of a Threshold
+value and a Randomness value, each containing 56 bits of information.
+
+A constant known as _maximum adjusted count_ (`MaxAdjustedCount`),
```
It could be just me, but I think that `max`-_something_ suggests inclusiveness, so this can be confusing. How about `AdjustedCountLimit`?

oertl replied:
I am fine with `MaxAdjustedCount`. It is inclusive with respect to the adjusted count. However, I understand that it can be a little confusing as it is also used as an exclusive upper limit for the threshold and the random value.

PeterF778 commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,262 @@
+probability, and `MaxAdjustedCount` (defined above) is the inverse of
+the smallest supported sampling probability (which can also be
+represented as `0x1p-56`, `math.Pow(2, -56)`, or `math.Ldexp(1,
+-56)`).
```
The above is a bit inaccurate (the smallest supported sampling probability is actually 1/(MaxAdjustedCount-1), I suggest removing the second part of the sentence.

oertl replied:
The smallest sampling probability is `2^-56`, which is realized with a threshold of `2^56-1 = MaxAdjustedCount - 1`. So I think the proposal is correct.

PeterF778 replied:
You are right. I cannot count correctly ...

kentquirk commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,262 @@
+### Sampling threshold
+
+When determining the Threshold value from an item of telemetry,
+sampler implementations SHOULD:
```
```suggestion
sampler implementations SHOULD evaluate the following in order:
```
 
Or some other wording that indicates that the first value found should be used.

kentquirk commented on docs/sampling/README.md:
```diff
@@ -0,0 +1,262 @@
+### Sampling randomness
+
+When determining the Randomness value from an item of telemetry,
+sampler implementations SHOULD:
```
```suggestion
sampler implementations SHOULD evaluate the following in order:
```

kentquirk reviewed (approved):
A minor wording clarification, but otherwise this looks good.

lmolkova commented:
@jmacd do we have a consensus on this PR within the sampling WG?

jmacd commented:
Miscellaneous updates:

Apologies for stalling. I am re-opening this with the intention to start it moving again, next week. 

Related: @kalyanaj is working on minor revisions to OTEP 235 based on our analysis of the W3C tracestate level 2 "random" flag.

Related: The semantic conventions here go slightly beyond OTEP 235, which focuses on tracestate and does not explicitly document our approach to logs sampling. The new semantic conventions in this PR _could be_ added back into OTEP 235 as there is no disagreement in the Sampling SIG, or we could just approve them here -- I am referring to the new `sampling.randomness` and `sampling.threshold` attributes.

The OTel Collector prototype for this is in review in its final stage. If we don't get thus work approved and merged soon, the work done there will be at-risk for near-future breaking changes. Please see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31894.
